subsys/shell/shell.c

```diff
@@ -1331,25 +1333,57 @@ void shell_fprintf(const struct shell *shell, enum shell_vt100_color color,
 	k_mutex_unlock(&shell->ctx->wr_mtx);
 }
 
-void shell_hexdump(const struct shell *shell, const u8_t *data, size_t len)
+static void shell_hexdump_line(const struct shell *shell, unsigned int offset,
+			       const u8_t *data, size_t len)
 {
-	int n = 0;
+	int i;
+
+	shell_fprintf(shell, SHELL_NORMAL, "%08X: ", offset);
 
-	while (len--) {
-		if (n % 16 == 0) {
-			shell_fprintf(shell, SHELL_NORMAL, "%08X: ", n);
+	for (i = 0; i < HEXDUMP_BYTES_IN_LINE; i++) {
+		if (i > 0 && !(i % 8)) {
+			shell_fprintf(shell, SHELL_NORMAL, " ");
 		}
 
-		shell_fprintf(shell, SHELL_NORMAL, "%02X ", *data++);
+		if (i < len) {
+			shell_fprintf(shell, SHELL_NORMAL, "%02x ",
+				      data[i] & 0xFF);
+		} else {
+			shell_fprintf(shell, SHELL_NORMAL, "   ");
+		}
+	}
+
+	shell_fprintf(shell, SHELL_NORMAL, "|");
 
-		n++;
-		if (n % 16 == 0) {
-			shell_print(shell, "");
+	for (i = 0; i < HEXDUMP_BYTES_IN_LINE; i++) {
+		if (i > 0 && !(i % 8)) {
+			shell_fprintf(shell, SHELL_NORMAL, " ");
+		}
+
+		if (i < len) {
+			char c = data[i];
+
+			shell_fprintf(shell, SHELL_NORMAL, "%c",
+				      isprint((int)c) ? c : '.');
+		} else {
+			shell_fprintf(shell, SHELL_NORMAL, " ");
 		}
 	}
+	shell_print(shell, "");
```

Comment:
It might be useful to terminate the printable section with a '|' to make sure whitespaces at the end of the line are visible.

Reply:
I agree with that.